Application can reset stream or close connection even if no error occurs. (#3953)
quicwg/base-drafts: draft-ietf-quic-http.md

# Error Handling {#errors}

QUIC allows the application to abruptly terminate (reset) individual streams or

> the entire connection, see Sections 2.4 and 5.3 of {{QUIC-TRANSPORT}}.  The
> cause for this is referred to as a "stream error" or "connection error" (see

Reviewer: ```suggestion
the entire connection; see Sections 2.4 and 5.3 of {{QUIC-TRANSPORT}}.  This
is referred to as a "stream error" or "connection error" (see
```

> {{QUIC-TRANSPORT}}) and have an associated error code, but do not necessarily

Reviewer 2: ```suggestion
{{QUIC-TRANSPORT}}) and have associated error codes, but do not necessarily
```

> necessarily have to be an error.  For example, a stream can be reset if the

Reviewer 2: This phrasing is strange. What does it mean to not "be an error" when these are referred to as "stream error" or "connection error" ?

Author: I agree with you.  My interpretation is that the source of strangeness is calling these stream errors and connection errors in the first place, and calling their reason codes error codes.  This way "error" becomes a technical term defined in this document, which does not align with my mental model of what an error is: entry found is cache or user closing a tab or navigating away is clearly not an error, but the specs treat it is a stream error and assign an error code to it.

Unfortunately I cannot think of a less weird way of phrasing it.  "Do not necessarily have to be abnormal events or protocol violations."?

Reviewer 3: While I agree that it sounds a bit strange, I think the current text is no more confusing as the NO_ERROR error code that we have in the transport draft.

> have to be errors.  For example, a stream can be reset if the requested resource
> is no longer needed.

Reviewer: Given the comment thread, what about....
```suggestion
indicate a problem with the connection or either implementation.  For example, a
stream can be reset if the requested resource is no longer needed.
```

Author: Thank you, I like this wording.

> the entire connection; see Sections 2.4 and 5.3 of {{QUIC-TRANSPORT}}.  This
> is referred to as a "stream error" or "connection error" (see
> Section 11 of {{QUIC-TRANSPORT}}) and has an associated error code, but does not
> necessarily have to be an error.  For example, a stream can be reset if the
> requested resource is no longer needed.

Reviewer 2: ```suggestion
the entire connection; see Sections 2.4 and 5.3 of {{QUIC-TRANSPORT}}.  These
are referred to as a "stream error" or "connection error" (Section 11 of
{{QUIC-TRANSPORT}}) and have an associated error code, but do not
necessarily have to be errors.  For example, a stream can be reset if the
requested resource is no longer needed.
```

Author: I'm accepting your suggestion but also pluralizing "stream error" and "connection error" (which incidentally were already plural before this PR).
